In [2]:
import pandas as pd

#loading train datasets
train_customers = pd.read_csv('../data/raw/train/df_Customers.csv')
train_orderitems = pd.read_csv('../data/raw/train/df_OrderItems.csv')
train_orders = pd.read_csv('../data/raw/train/df_Orders.csv')
train_payments = pd.read_csv('../data/raw/train/df_Payments.csv')
train_products = pd.read_csv('../data/raw/train/df_Products.csv')

#loading test datasets
test_customers = pd.read_csv('../data/raw/test/df_Customers.csv')
test_orderitems = pd.read_csv('../data/raw/test/df_OrderItems.csv')
test_orders = pd.read_csv('../data/raw/test/df_Orders.csv')
test_payments = pd.read_csv('../data/raw/test/df_Payments.csv')
test_products = pd.read_csv('../data/raw/test/df_Products.csv')

print("All 10 files are loaded\n")


All 10 files are loaded



In [4]:
import os
print(f"Notebook saved at: {os.getcwd()}")
print(f"Files in this folder: {os.listdir()}")

Notebook saved at: C:\Users\User\Desktop\Miuni\orderiq_project
Files in this folder: ['.ipynb_checkpoints', 'Ecommerce Order Dataset', 'orderiq.ipynb']


In [5]:
#Exploring datasets
print("Customers")
print("="*60)
df1 = train_customers #Remember to change this if you need to know abt other tables

print("SHAPE: ", df1.shape)
print("\nCOLUMNS: ", df1.columns.tolist())
print("\nFIRST FIVE ROWS: ", df1.head())
print("\nDATA TYPES: ", df1.dtypes)
print("\nMISSING VALUES: ", df1.isnull().sum())
print("\nBASIC STATISTICS:", df1.describe())

print("Order Items")
print("="*60)
df2 = train_orderitems #Remember to change this if you need to know abt other tables

print("SHAPE: ", df2.shape)
print("\nCOLUMNS: ", df2.columns.tolist())
print("\nFIRST FIVE ROWS: ", df2.head())
print("\nDATA TYPES: ", df2.dtypes)
print("\nMISSING VALUES: ", df2.isnull().sum())
print("\nBASIC STATISTICS:", df2.describe())

print("Orders")
print("="*60)
df3 = train_orders #Remember to change this if you need to know abt other tables

print("SHAPE: ", df3.shape)
print("\nCOLUMNS: ", df3.columns.tolist())
print("\nFIRST FIVE ROWS: ", df3.head())
print("\nDATA TYPES: ", df3.dtypes)
print("\nMISSING VALUES: ", df3.isnull().sum())
print("\nBASIC STATISTICS:", df3.describe())

print("Payments")
print("="*60)
df4 = train_payments#Remember to change this if you need to know abt other tables

print("SHAPE: ", df4.shape)
print("\nCOLUMNS: ", df4.columns.tolist())
print("\nFIRST FIVE ROWS: ", df4.head())
print("\nDATA TYPES: ", df4.dtypes)
print("\nMISSING VALUES: ", df4.isnull().sum())
print("\nBASIC STATISTICS:", df4.describe())

print("Products")
print("="*60)
df5 = train_products #Remember to change this if you need to know abt other tables

print("SHAPE: ", df5.shape)
print("\nCOLUMNS: ", df5.columns.tolist())
print("\nFIRST FIVE ROWS: ", df5.head())
print("\nDATA TYPES: ", df5.dtypes)
print("\nMISSING VALUES: ", df5.isnull().sum())
print("\nBASIC STATISTICS:", df5.describe())






Customers
SHAPE:  (89316, 4)

COLUMNS:  ['customer_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

FIRST FIVE ROWS:      customer_id  customer_zip_code_prefix       customer_city customer_state
0  hCT0x9JiGXBQ                     58125     varzea paulista             SP
1  PxA7fv9spyhx                      3112  armacao dos buzios             RJ
2  g3nXeJkGI0Qw                      4119             jandira             SP
3  EOEsCQ6QlpIg                     18212          uberlandia             MG
4  mVz5LO2Vd6cL                     88868            ilhabela             SP

DATA TYPES:  customer_id                 object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

MISSING VALUES:  customer_id                 0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

BASIC STATISTICS:        customer_zip_code_prefix
count              89316.00000

*train_customers*
SHAPE:  (89316, 4)

COLUMNS:  ['customer_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

FIRST FIVE ROWS:      customer_id  customer_zip_code_prefix       customer_city customer_state
0  hCT0x9JiGXBQ                     58125     varzea paulista             SP
1  PxA7fv9spyhx                      3112  armacao dos buzios             RJ
2  g3nXeJkGI0Qw                      4119             jandira             SP
3  EOEsCQ6QlpIg                     18212          uberlandia             MG
4  mVz5LO2Vd6cL                     88868            ilhabela             SP

DATA TYPES:  customer_id                 object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

MISSING VALUES:  customer_id                 0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

BASIC STATISTICS:        customer_zip_code_prefix
count              89316.000000
mean               40499.471080
std                31194.386361
min                 1003.000000
25%                 9785.000000
50%                35480.000000
75%                70040.000000
max                99990.000000
Selection deleted


*train_orderitems*
SHAPE:  (89316, 5)

COLUMNS:  ['order_id', 'product_id', 'seller_id', 'price', 'shipping_charges']

FIRST FIVE ROWS:         order_id    product_id     seller_id   price  shipping_charges
0  Axfy13Hk4PIk  90K0C1fIyQUf  ZWM05J9LcBSF  223.51             84.65
1  v6px92oS8cLG  qejhpMGGVcsl  IjlpYfhUbRQs  170.80             23.79
2  Ulpf9skrhjfm  qUS5d2pEAyxJ  77p2EYxcM9MD   64.40             17.38
3  bwJVWupf2keN  639iGvMyv0De  jWzS0ayv9TGf  264.50             30.72
4  Dd0QnrMk9Cj5  1lycYGcsic2F  l1pYW6GBnPMr  779.90             30.66

DATA TYPES:  order_id             object
product_id           object
seller_id            object
price               float64
shipping_charges    float64
dtype: object

MISSING VALUES:  order_id            0
product_id          0
seller_id           0
price               0
shipping_charges    0
dtype: int64

BASIC STATISTICS:               price  shipping_charges
count  89316.000000      89316.000000
mean     340.900543         44.283210
std      557.459897         37.672491
min        0.850000          0.000000
25%       59.650000         20.110000
50%      136.900000         35.055000
75%      399.200000         57.190000
max     6735.000000        409.680000

*train_orders*
SHAPE:  (89316, 7)

COLUMNS:  ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_timestamp', 'order_estimated_delivery_date']

FIRST FIVE ROWS:         order_id   customer_id order_status order_purchase_timestamp  \
0  Axfy13Hk4PIk  hCT0x9JiGXBQ    delivered      2017-10-22 18:57:54   
1  v6px92oS8cLG  PxA7fv9spyhx    delivered      2018-06-20 21:40:31   
2  Ulpf9skrhjfm  g3nXeJkGI0Qw    delivered      2018-02-16 16:19:31   
3  bwJVWupf2keN  EOEsCQ6QlpIg    delivered      2018-08-18 18:04:29   
4  Dd0QnrMk9Cj5  mVz5LO2Vd6cL    delivered      2017-12-22 16:44:04   

     order_approved_at order_delivered_timestamp order_estimated_delivery_date  
0  2017-10-22 19:14:13       2017-10-26 22:19:52                    2017-11-09  
1  2018-06-20 22:20:20       2018-07-03 22:51:22                    2018-07-24  
2  2018-02-17 16:15:35       2018-02-27 01:29:50                    2018-03-08  
3  2018-08-18 18:15:16       2018-08-27 20:03:51                    2018-09-19  
4  2017-12-22 17:31:31       2018-01-05 19:22:49                    2018-01-18  

DATA TYPES:  order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_timestamp        object
order_estimated_delivery_date    object
dtype: object

MISSING VALUES:  order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   9
order_delivered_timestamp        1889
order_estimated_delivery_date       0
dtype: int64

BASIC STATISTICS:             order_id   customer_id order_status order_purchase_timestamp  \
count          89316         89316        89316                    89316   
unique         89316         89316            7                    72775   
top     Axfy13Hk4PIk  hCT0x9JiGXBQ    delivered      2017-08-08 20:26:31   
freq               1             1        87428                       87   

          order_approved_at order_delivered_timestamp  \
count                 89307                     87427   
unique                68215                     71143   
top     2017-08-08 20:43:31       2017-08-14 12:46:18   
freq                     87                        87   

       order_estimated_delivery_date  
count                          89316  
unique                           444  
top                       2017-12-20  

*train_payments*
SHAPE:  (89316, 5)

COLUMNS:  ['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value']

FIRST FIVE ROWS:         order_id  payment_sequential payment_type  payment_installments  \
0  Axfy13Hk4PIk                   1  credit_card                     1   
1  v6px92oS8cLG                   1  credit_card                     8   
2  Ulpf9skrhjfm                   1  credit_card                     4   
3  bwJVWupf2keN                   1  credit_card                     2   
4  Dd0QnrMk9Cj5                   1  credit_card                     1   

   payment_value  
0         259.14  
1         382.39  
2         249.25  
3          27.79  
4          76.15  

DATA TYPES:  order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

MISSING VALUES:  order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

BASIC STATISTICS:        payment_sequential  payment_installments  payment_value
count        89316.000000          89316.000000   89316.000000
mean             1.099646              2.965717     268.657190
std              0.772658              2.796406     344.409566
min              1.000000              0.000000       0.000000
25%              1.000000              1.000000      84.340000
50%              1.000000              2.000000     171.860000
75%              1.000000              4.000000     313.530000
max             29.000000             24.000000    7274.880000

*train_products*
SHAPE:  (89316, 6)

COLUMNS:  ['product_id', 'product_category_name', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']

FIRST FIVE ROWS:       product_id     product_category_name  product_weight_g  \
0  90K0C1fIyQUf                      toys             491.0   
1  qejhpMGGVcsl             watches_gifts             440.0   
2  qUS5d2pEAyxJ  costruction_tools_garden            2200.0   
3  639iGvMyv0De                      toys            1450.0   
4  1lycYGcsic2F                      toys             300.0   

   product_length_cm  product_height_cm  product_width_cm  
0               19.0               12.0              16.0  
1               18.0               14.0              17.0  
2               16.0               16.0              16.0  
3               68.0                3.0              48.0  
4               17.0                4.0              12.0  

DATA TYPES:  product_id                object
product_category_name     object
product_weight_g         float64
product_length_cm        float64
product_height_cm        float64
product_width_cm         float64
dtype: object

MISSING VALUES:  product_id                 0
product_category_name    308
product_weight_g          15
product_length_cm         15
product_height_cm         15
product_width_cm          15
dtype: int64

BASIC STATISTICS:        product_weight_g  product_length_cm  product_height_cm  \
count      89301.000000       89301.000000       89301.000000   
mean        2087.068129          30.220658          16.559423   
std         3747.039215          16.110199          13.388514   
min            0.000000           7.000000           2.000000   
25%          300.000000          18.000000           8.000000   
50%          700.000000          25.000000          13.000000   
75%         1800.000000          38.000000          20.000000   
max        40425.000000         105.000000         105.000000   

       product_width_cm  
count       89301.00000  
mean           23.03421  
std            11.68828  
min             6.00000  
25%            15.00000  
50%            20.00000  
75%            30.00000  
max           118.00000  
Selection deleted



In [6]:
print("CLEANING DATA")
print("="*60)
#checking for duplicates

#Check Customers - train
print("Customers\n")
print(f"Total Rows(Before):{len(train_customers)}")
duplicate_count = train_customers.duplicated(subset='customer_id').sum() #Duplicated Rows Count
print(f"Duplicated Rows Count:{train_customers}")
train_customers= train_customers.drop_duplicates(subset='customer_id', keep='first')
print(f"Total Rows(After):{len(train_customers)}")

#Check Orders - train
print("Orders\n")
print(f"Total Rows(Before):{len(train_orders)}")
duplicate_count = train_orders.duplicated(subset='order_id').sum() #Duplicated Rows Count
print(f"Duplicated Rows Count:{train_orders}")
train_orders= train_orders.drop_duplicates(subset='order_id', keep='first')
print(f"Total Rows(After):{len(train_orders)}")

#Check Products - train
print("Products\n")
print(f"Total Rows(Before):{len(train_products)}")
duplicate_count = train_products.duplicated(subset='product_id').sum() #Duplicated Rows Count
print(f"Duplicated Rows Count:{train_products}")
train_products= train_products.drop_duplicates(subset='product_id', keep='first')
print(f"Total Rows(After):{len(train_products)}")



CLEANING DATA
Customers

Total Rows(Before):89316
Duplicated Rows Count:        customer_id  customer_zip_code_prefix       customer_city  \
0      hCT0x9JiGXBQ                     58125     varzea paulista   
1      PxA7fv9spyhx                      3112  armacao dos buzios   
2      g3nXeJkGI0Qw                      4119             jandira   
3      EOEsCQ6QlpIg                     18212          uberlandia   
4      mVz5LO2Vd6cL                     88868            ilhabela   
...             ...                       ...                 ...   
89311  bfJtzkBGksdY                     94060              amparo   
89312  4hjsCpj6FC7g                     99835             itapevi   
89313  DLn56oKHl9bv                      2540           sao paulo   
89314  auSfRYUQUOIq                     12288            paulinia   
89315  rhevVNXYJnEs                     93048            brasilia   

      customer_state  
0                 SP  
1                 RJ  
2                 SP  
3      

In [7]:
print("HANDLING MISSING VALUES")
print("="*60)
print(f"Customer Table missing values:{train_customers.isnull().sum()}")#Customer Table
print(f"Order Table missing values:{train_orders.isnull().sum()}")#Order Table
train_orders['order_approved_at'] = train_orders['order_approved_at'].fillna(
    train_orders['order_purchase_timestamp'])
print("Orders cleaned\n")
print(f"Product Table missing values:{train_products.isnull().sum()}")#Product Table
train_products.loc[:,'product_category_name'] = train_products['product_category_name'].fillna('Unknown')
num_cols = ['product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']
train_products.loc[:,num_cols] = train_products[num_cols].fillna(train_products[num_cols].median())
print("Products cleaned\n")

HANDLING MISSING VALUES
Customer Table missing values:customer_id                 0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
Order Table missing values:order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   9
order_delivered_timestamp        1889
order_estimated_delivery_date       0
dtype: int64
Orders cleaned

Product Table missing values:product_id                 0
product_category_name    141
product_weight_g           2
product_length_cm          2
product_height_cm          2
product_width_cm           2
dtype: int64
Products cleaned



In [13]:
print("DATA CLEANING - CHECKING DATA TYPES")
print("="*60)
#Converting date columns into datetime
date_columns = ['order_purchase_timestamp','order_approved_at','order_delivered_timestamp']
for column in date_columns:
    train_orders[column]= pd.to_datetime(train_orders[column],errors='coerce')
    print("Converted {column} into datetime")
train_orders['order_estimated_delivery_date']=pd.to_datetime(train_orders['order_estimated_delivery_date'], errors='coerce').dt.date
print("Converted to date")
#Check numerical columns & make zero as the lowest
numerical_cols=['price','shipping_charges','payment_value']
for col in numerical_cols:
    if col in train_orderitems.columns:
        print(f"\n{col} - Min:{train_orderitems[col].min()}, Max:{train_orderitems[col].max()}")
        if(train_orderitems[col]<0).any():
            print(f"Found Negative values in{col}")
            train_orderitems[col]=train_orderitems[col].clip(lower=0)
            print("Fixed Negative Values")


DATA CLEANING - CHECKING DATA TYPES
Converted {column} into datetime
Converted {column} into datetime
Converted {column} into datetime
Converted to date

price - Min:0.85, Max:6735.0

shipping_charges - Min:0.0, Max:409.68


In [22]:
print("DATA CLEANING - VALIDATING RELATIONSHIPS")
print("="*60)
valid_orders = set(train_orders['order_id'])#Checking all orders are valid
order_items_orders = set(train_orderitems['order_id'])
missing_orders = order_items_orders - valid_orders
print(f"Order items with valid orders: {len(order_items_orders & valid_orders)}/{len(order_items_orders)}")
if missing_orders:
    print(f"Found {len(missing_orders)} order items with invalid IDs")
    train_orderitems = train_orderitems[train_orderitems['order_id'].isin(valid_orders)]
    print("Removed invalid order items")
#checking all order items have valid products
valid_products = set(train_products['product_id'])
order_items_products = set(train_products['product_id'])
missing_products = order_items_products - valid_products
print(f"Order items with valid products: {len(order_items_products & valid_products)}/{len(order_items_products)}")
if missing_products:
    print(f"Found {len(missing_products)} order items with invalid product IDs")
    train_orderitems = train_orderitems[train_orderitems['product_id'].isin(valid_products)]
    print("Removed invalid order items")
#Checking all orders have valid customers
valid_customers = set(train_customers['customer_id'])
orders_customers = set(train_orders['customer_id'])
missing_customers = orders_customers - valid_customers
print(f"Order items with valid customers: {len(orders_customers & valid_customers)}/{len(orders_customers)}")
if missing_customers:
    print(f"Found {len(missing_customers)} order items with invalid customer IDs")
    train_orders = train_orders[train_orders['customer_id'].isin(valid_customers)]
    print("Removed invalid orders")

DATA CLEANING - VALIDATING RELATIONSHIPS
Order items with valid orders: 89316/89316
Order items with valid products: 27451/27451
Order items with valid customers: 89316/89316


In [23]:
#for backup
train_customers.to_csv('Ecommerce Order Dataset/train/df_Customers_cleaned.csv', index=False)
train_orderitems.to_csv('Ecommerce Order Dataset/train/df_Orderitems_cleaned.csv', index=False)
train_orders.to_csv('Ecommerce Order Dataset/train/df_Orders_cleaned.csv', index=False)
train_products.to_csv('Ecommerce Order Dataset/train/df_Products_cleaned.csv', index=False)
train_payments.to_csv('Ecommerce Order Dataset/train/df_Payments_cleaned.csv', index=False)

In [25]:
import pyodbc
import pandas as pd
from datetime import datetime

# Connect
conn = pyodbc.connect('DRIVER={ODBC Driver 17 for SQL Server};SERVER=LAPTOP-8G4F56T7\\SQLEXPRESS;DATABASE=OrderIqDB;Trusted_Connection=yes')
cursor = conn.cursor()

# Function to safely convert dates
def safe_date(date_val):
    if pd.isna(date_val) or date_val == '1900-01-01' or date_val == '':
        return None
    try:
        return pd.to_datetime(date_val)
    except:
        return None

# Insert customers (this works fine)
print("Inserting customers...")
for _, row in train_customers.iterrows():
    cursor.execute("INSERT INTO customers VALUES (?,?,?,?)", 
                   row.customer_id, row.customer_zip_code_prefix, 
                   row.customer_city, row.customer_state)
print(f"Inserted {len(train_customers)} customers")

# Insert products
print("Inserting products...")
for _, row in train_products.iterrows():
    cursor.execute("INSERT INTO products VALUES (?,?,?,?,?,?)", 
                   row.product_id, row.product_category_name, row.product_weight_g,
                   row.product_length_cm, row.product_height_cm, row.product_width_cm)
print(f"Inserted {len(train_products)} products")

# Insert orders with safe date conversion
print("Inserting orders...")
for _, row in train_orders.iterrows():
    # Convert dates safely
    purchase_date = safe_date(row.order_purchase_timestamp)
    approved_date = safe_date(row.order_approved_at)
    delivered_date = safe_date(row.order_delivered_timestamp)
    estimated_date = safe_date(row.order_estimated_delivery_date)
    
    cursor.execute("INSERT INTO orders VALUES (?,?,?,?,?,?,?)",
                   row.order_id, row.customer_id, row.order_status,
                   purchase_date, approved_date, delivered_date, estimated_date)
print(f"Inserted {len(train_orders)} orders")

conn.commit()
print("\nData inserted successfully!")

Inserting customers...
Inserted 89316 customers
Inserting products...
Inserted 27451 products
Inserting orders...
Inserted 89316 orders

Data inserted successfully!


In [3]:
import pandas as pd
from sqlalchemy import create_engine 
engine = create_engine(
    "mssql+pyodbc://LAPTOP-8G4F56T7\\SQLEXPRESS/OrderIqDB?driver=ODBC+Driver+17+for+SQL+Server"
)
#Query
query = """
SELECT 
    o.order_id,
    o.order_status,
    o.order_purchase_timestamp,
    o.order_approved_at,
    o.order_delivered_timestamp,
    o.order_estimated_delivery_date,
    
    c.customer_id,
    c.customer_city,
    c.customer_state,
    
    -- AGGREGATED PRODUCT INFO
    COUNT(DISTINCT oi.product_id) AS total_products,
    -- Alternative to STRING_AGG for SQL Server
    STUFF(
        (SELECT DISTINCT ', ' + p2.product_category_name
         FROM order_items oi2 
         JOIN products p2 ON oi2.product_id = p2.product_id
         WHERE oi2.order_id = o.order_id
         FOR XML PATH('')), 1, 2, '') AS product_categories,
    SUM(oi.price) AS total_price,
    SUM(oi.shipping_charges) AS total_shipping,
    
    -- AGGREGATED PAYMENT INFO
    COUNT(DISTINCT pymt.payment_sequential) AS payment_count,
    STUFF(
        (SELECT DISTINCT ', ' + p2.payment_type
         FROM payments p2 
         WHERE p2.order_id = o.order_id
         FOR XML PATH('')), 1, 2, '') AS payment_types,
    SUM(pymt.payment_value) AS total_payment_value,
    MAX(pymt.payment_installments) AS max_installments,
    
    -- DERIVED FIELDS
    DATEDIFF(DAY, o.order_purchase_timestamp, o.order_delivered_timestamp) AS delivery_time_days,
    DATEDIFF(DAY, o.order_purchase_timestamp, o.order_estimated_delivery_date) AS estimated_delivery_days,
    CASE 
        WHEN o.order_delivered_timestamp > o.order_estimated_delivery_date THEN 1 
        WHEN o.order_delivered_timestamp IS NULL THEN NULL
        ELSE 0 
    END AS is_late

FROM orders o
    LEFT JOIN customers c ON o.customer_id = c.customer_id
    LEFT JOIN order_items oi ON o.order_id = oi.order_id
    LEFT JOIN products p ON oi.product_id = p.product_id
    LEFT JOIN payments pymt ON o.order_id = pymt.order_id

GROUP BY 
    o.order_id,
    o.order_status,
    o.order_purchase_timestamp,
    o.order_approved_at,
    o.order_delivered_timestamp,
    o.order_estimated_delivery_date,
    c.customer_id,
    c.customer_city,
    c.customer_state
"""

print("Creating Master Dataset")
master_df = pd.read_sql(query, engine)
print(f"Rows: {master_df.shape[0]}, Columns: {master_df.shape[1]}")
print(master_df.head())

print("\nColumn Names:")
print(master_df.columns)

print("\nMissing Values:")
print(master_df.isnull().sum())

Creating Master Dataset
Rows: 89316, Columns: 20
       order_id order_status order_purchase_timestamp   order_approved_at  \
0  001gkk6BwKEB    delivered      2018-04-16 14:01:44 2018-04-16 14:32:36   
1  0029bAIZFMKA    delivered      2018-02-17 15:04:06 2018-02-20 07:06:53   
2  002Jk8hp9Bhy    delivered      2018-04-24 09:27:39 2018-04-25 03:51:36   
3  002ohHsWSKrP    delivered      2018-05-17 17:09:33 2018-05-17 20:32:27   
4  0041NijdAYB0    delivered      2018-07-10 10:32:53 2018-07-10 10:45:39   

  order_delivered_timestamp order_estimated_delivery_date   customer_id  \
0       2018-04-20 19:42:15                    2018-05-03  xKv2nBFnvcwS   
1       2018-03-22 23:51:19                    2018-03-15  Xcm6u682SuQQ   
2       2018-04-26 18:17:28                    2018-05-17  l0HoYdeR1BJM   
3       2018-05-21 20:31:47                    2018-05-25  z5refCsXE1nS   
4       2018-07-25 12:36:45                    2018-08-01  XMxGRoDyISTv   

     customer_city customer_state  to

In [3]:
test_query = """
SELECT TOP 5 o.order_id, COUNT(oi.product_id) AS product_count
FROM orders o
LEFT JOIN order_items oi ON o.order_id = oi.order_id
GROUP BY o.order_id
"""
test_df = pd.read_sql(test_query, engine)
print(test_df)

       order_id  product_count
0  001gkk6BwKEB              0
1  0029bAIZFMKA              0
2  002Jk8hp9Bhy              0
3  002ohHsWSKrP              0
4  0041NijdAYB0              0


In [4]:
# Check row count in order_items
items_count = pd.read_sql("SELECT COUNT(*) AS total_rows FROM order_items", engine)
print(f"Order items rows: {items_count['total_rows'].iloc[0]}")

# Check a sample of what's in order_items
items_sample = pd.read_sql("SELECT TOP 5 * FROM order_items", engine)
print("\nOrder items sample:")
print(items_sample)

# Check a sample of orders
orders_sample = pd.read_sql("SELECT TOP 5 order_id FROM orders", engine)
print("\nOrders sample:")
print(orders_sample)

Order items rows: 0

Order items sample:
Empty DataFrame
Columns: [order_id, product_id, seller_id, price, shipping_charges]
Index: []

Orders sample:
       order_id
0  001gkk6BwKEB
1  0029bAIZFMKA
2  002Jk8hp9Bhy
3  002ohHsWSKrP
4  0041NijdAYB0


In [6]:
import pyodbc
import pandas as pd

conn = pyodbc.connect('DRIVER={ODBC Driver 17 for SQL Server};SERVER=LAPTOP-8G4F56T7\\SQLEXPRESS;DATABASE=OrderIqDB;Trusted_Connection=yes')
cursor = conn.cursor()

# Load your original data files - UPDATE THESE PATHS to where your files are
train_orderitems = pd.read_csv('C:/Users/User/Desktop/Miuni/orderiq_project/Ecommerce Order Dataset/train/df_OrderItems.csv')
train_payments = pd.read_csv('C:/Users/User/Desktop/Miuni/orderiq_project/Ecommerce Order Dataset/train/df_Payments.csv')

# ============================================
# INSERT ORDER ITEMS
# ============================================
print("Inserting order_items...")
count = 0
for _, row in train_orderitems.iterrows():
    cursor.execute("""
        INSERT INTO order_items (order_id, product_id, seller_id, price, shipping_charges) 
        VALUES (?, ?, ?, ?, ?)
    """, row.order_id, row.product_id, row.seller_id, row.price, row.shipping_charges)
    count += 1
    if count % 10000 == 0:
        print(f"  Inserted {count} rows...")
        conn.commit()

conn.commit()
print(f"Inserted {count} order_items rows")

# ============================================
# INSERT PAYMENTS
# ============================================
print("\nInserting payments...")
count = 0
for _, row in train_payments.iterrows():
    cursor.execute("""
        INSERT INTO payments (order_id, payment_sequential, payment_type, payment_installments, payment_value) 
        VALUES (?, ?, ?, ?, ?)
    """, row.order_id, row.payment_sequential, row.payment_type, 
       row.payment_installments, row.payment_value)
    count += 1
    if count % 10000 == 0:
        print(f"  Inserted {count} rows...")
        conn.commit()

conn.commit()
print(f"Inserted {count} payments rows")

# ============================================
# VERIFY THE DATA
# ============================================
print("\n" + "="*50)
print("VERIFYING DATA...")
print("="*50)

verify = pd.read_sql("""
    SELECT 
        (SELECT COUNT(*) FROM order_items) AS order_items_count,
        (SELECT COUNT(*) FROM payments) AS payments_count,
        (SELECT COUNT(DISTINCT order_id) FROM order_items) AS distinct_orders_in_items,
        (SELECT COUNT(DISTINCT order_id) FROM payments) AS distinct_orders_in_payments,
        (SELECT COUNT(*) FROM orders) AS total_orders
""", conn)

print(verify)

cursor.close()
conn.close()
print("\n All data inserted successfully!")

Inserting order_items...
  Inserted 10000 rows...
  Inserted 20000 rows...
  Inserted 30000 rows...
  Inserted 40000 rows...
  Inserted 50000 rows...
  Inserted 60000 rows...
  Inserted 70000 rows...
  Inserted 80000 rows...
Inserted 89316 order_items rows

Inserting payments...
  Inserted 10000 rows...
  Inserted 20000 rows...
  Inserted 30000 rows...
  Inserted 40000 rows...
  Inserted 50000 rows...
  Inserted 60000 rows...
  Inserted 70000 rows...
  Inserted 80000 rows...
Inserted 89316 payments rows

VERIFYING DATA...


C:\Users\User\AppData\Local\Temp\ipykernel_48648\420230622.py:55: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  verify = pd.read_sql("""


   order_items_count  payments_count  distinct_orders_in_items  \
0              89316           89316                     89316   

   distinct_orders_in_payments  total_orders  
0                        89316         89316  

✅ All data inserted successfully!


In [7]:
# Now run the master query
master_df = pd.read_sql(query, engine)
print(f"Rows: {master_df.shape[0]}")
print("\nFirst order with products:")
print(master_df[master_df['total_products'] > 0][['order_id', 'total_products', 'product_categories', 'total_price']].head())

Rows: 89316

First order with products:
       order_id  total_products product_categories  total_price
0  001gkk6BwKEB               1               toys        12.69
1  0029bAIZFMKA               1          telephony       692.00
2  002Jk8hp9Bhy               1               toys        41.79
3  002ohHsWSKrP               1               toys       459.00
4  0041NijdAYB0               1               toys       518.18


In [8]:
# ============================================
# FEATURE ENGINEERING - DELIVERY FEATURES
# ============================================

# Convert to datetime (already done in query, but safe to ensure)
master_df['order_purchase_timestamp'] = pd.to_datetime(master_df['order_purchase_timestamp'])
master_df['order_delivered_timestamp'] = pd.to_datetime(master_df['order_delivered_timestamp'])
master_df['order_estimated_delivery_date'] = pd.to_datetime(master_df['order_estimated_delivery_date'])

# Delivery time in days (using our existing columns)
master_df['delivery_time'] = master_df['delivery_time_days']  # Already calculated in SQL

# Estimated delivery time in days
master_df['estimated_delivery_time'] = master_df['estimated_delivery_days']  # Already calculated

# Delivery delay (positive = late, negative = early)
master_df['delivery_delay'] = master_df['delivery_time_days'] - master_df['estimated_delivery_days']

# ============================================
# FEATURE ENGINEERING - ORDER VALUE FEATURES
# ============================================

# Total order value (using our aggregated columns)
master_df['total_order_value'] = master_df['total_price'] + master_df['total_shipping']

# Average item price (if multiple products)
master_df['avg_item_price'] = master_df['total_price'] / master_df['total_products']
master_df['avg_item_price'] = master_df['avg_item_price'].fillna(0)  # Handle orders with no products

# ============================================
# FEATURE ENGINEERING - TIME-BASED FEATURES
# ============================================

master_df['order_year'] = master_df['order_purchase_timestamp'].dt.year
master_df['order_month'] = master_df['order_purchase_timestamp'].dt.month
master_df['order_day'] = master_df['order_purchase_timestamp'].dt.day
master_df['order_hour'] = master_df['order_purchase_timestamp'].dt.hour

# Additional time features
master_df['order_day_of_week'] = master_df['order_purchase_timestamp'].dt.dayofweek  # 0=Monday
master_df['order_weekend'] = master_df['order_day_of_week'].apply(lambda x: 1 if x >= 5 else 0)
master_df['order_month_name'] = master_df['order_purchase_timestamp'].dt.month_name()

# ============================================
# FEATURE ENGINEERING - DELIVERY STATUS
# ============================================

master_df['is_delivered'] = master_df['order_status'].apply(lambda x: 1 if x == 'delivered' else 0)
master_df['is_late'] = master_df['delivery_delay'].apply(lambda x: 1 if x > 0 else 0)

# ============================================
# FEATURE ENGINEERING - CUSTOMER ORDER COUNT
# ============================================

master_df['customer_order_count'] = master_df.groupby('customer_id')['order_id'].transform('count')

# ============================================
# FEATURE ENGINEERING - PAYMENT FEATURES
# ============================================

# Average payment per installment
master_df['avg_payment_per_installment'] = master_df['total_payment_value'] / master_df['max_installments']
master_df['avg_payment_per_installment'] = master_df['avg_payment_per_installment'].fillna(master_df['total_payment_value'])

# ============================================
# FEATURE ENGINEERING - PRODUCT FEATURES
# ============================================

# Flag for multi-product orders
master_df['is_multi_product'] = (master_df['total_products'] > 1).astype(int)

# Shipping cost ratio
master_df['shipping_ratio'] = master_df['total_shipping'] / master_df['total_order_value']
master_df['shipping_ratio'] = master_df['shipping_ratio'].fillna(0)

# ============================================
# CLEAN UP - Handle Missing Values
# ============================================

# For undelivered orders, delivery-related fields should be None/0
master_df.loc[master_df['is_delivered'] == 0, 'delivery_time'] = None
master_df.loc[master_df['is_delivered'] == 0, 'delivery_delay'] = None
master_df.loc[master_df['is_delivered'] == 0, 'is_late'] = None

print("Feature engineering done!")
print(f"Final dataset shape: {master_df.shape}")
print(f"Columns: {master_df.columns.tolist()}")

Feature engineering done!
Final dataset shape: (89316, 37)
Columns: ['order_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_timestamp', 'order_estimated_delivery_date', 'customer_id', 'customer_city', 'customer_state', 'total_products', 'product_categories', 'total_price', 'total_shipping', 'payment_count', 'payment_types', 'total_payment_value', 'max_installments', 'delivery_time_days', 'estimated_delivery_days', 'is_late', 'delivery_time', 'estimated_delivery_time', 'delivery_delay', 'total_order_value', 'avg_item_price', 'order_year', 'order_month', 'order_day', 'order_hour', 'order_day_of_week', 'order_weekend', 'order_month_name', 'is_delivered', 'customer_order_count', 'avg_payment_per_installment', 'is_multi_product', 'shipping_ratio']


In [9]:
# ============================================
# CUSTOMER ANALYSIS
# ============================================
print("="*50)
print("CUSTOMER ANALYSIS")
print("="*50)

customer_analysis = master_df.groupby('customer_id').agg({
    'order_id': 'count',
    'total_payment_value': 'sum',  # Changed from 'payment_value'
    'total_order_value': 'sum',
    'customer_order_count': 'first'  # Already calculated
}).rename(columns={
    'order_id': 'total_orders',
    'total_payment_value': 'total_spent',
    'total_order_value': 'total_revenue'
})

print("\nTop 10 Customers by Total Spent:")
print(customer_analysis.nlargest(10, 'total_spent'))

# ============================================
# DELIVERY ANALYSIS
# ============================================
print("\n" + "="*50)
print("DELIVERY ANALYSIS")
print("="*50)

# Only analyze delivered orders
delivered_df = master_df[master_df['is_delivered'] == 1]

# On-time delivery rate (using our pre-calculated is_late)
on_time_rate = (delivered_df['is_late'] == 0).mean() * 100
print(f"On-time delivery rate: {on_time_rate:.2f}%")

# Average delivery time
avg_delivery = delivered_df['delivery_time'].mean()
print(f"Average delivery time: {avg_delivery:.1f} days")

# Late vs On-time comparison
late_vs_ontime = delivered_df.groupby('is_late').agg({
    'total_order_value': 'mean',
    'delivery_time': 'mean'
})
print("\nComparison (Late=1 vs On-time=0):")
print(late_vs_ontime)

# ============================================
# PRODUCT ANALYSIS
# ============================================
print("\n" + "="*50)
print("PRODUCT ANALYSIS")
print("="*50)

# Note: We need to explode product_categories for accurate analysis
# First, create a version with one row per product category
product_rows = []

for idx, row in master_df.iterrows():
    if row['product_categories'] and row['product_categories'] != '':
        categories = row['product_categories'].split(', ')
        for cat in categories:
            product_rows.append({
                'product_category_name': cat,
                'order_id': row['order_id'],
                'total_payment_value': row['total_payment_value'],
                'total_price': row['total_price']
            })

product_df = pd.DataFrame(product_rows)

if len(product_df) > 0:
    product_performance = product_df.groupby('product_category_name').agg({
        'order_id': 'count',
        'total_payment_value': 'sum'
    }).rename(columns={
        'order_id': 'total_orders',
        'total_payment_value': 'revenue'
    })
    
    print("\nTop 10 Product Categories by Revenue:")
    print(product_performance.nlargest(10, 'revenue'))
else:
    print("No product category data available")

# ============================================
# PAYMENT ANALYSIS
# ============================================
print("\n" + "="*50)
print("PAYMENT ANALYSIS")
print("="*50)

# Split payment_types into individual rows
payment_rows = []
for idx, row in master_df.iterrows():
    if row['payment_types'] and row['payment_types'] != '':
        payment_methods = row['payment_types'].split(', ')
        for method in payment_methods:
            payment_rows.append({
                'payment_type': method,
                'order_id': row['order_id'],
                'total_payment_value': row['total_payment_value']
            })

payment_df = pd.DataFrame(payment_rows)

if len(payment_df) > 0:
    print("\nPayment Type Distribution:")
    print(payment_df['payment_type'].value_counts())
    
    print("\nAverage Order Value by Payment Type:")
    print(payment_df.groupby('payment_type')['total_payment_value'].mean().sort_values(ascending=False))
else:
    print("No payment type data available")

# ============================================
# TIME TRENDS
# ============================================
print("\n" + "="*50)
print("TIME TRENDS")
print("="*50)

# Orders by month
monthly_orders = master_df.groupby('order_month')['order_id'].count()
print("\nMonthly Orders:")
print(monthly_orders)

# Orders by hour
hourly_orders = master_df.groupby('order_hour')['order_id'].count()
print("\nTop 5 Hours for Orders:")
print(hourly_orders.nlargest(5))

# Orders by day of week
weekly_orders = master_df.groupby('order_day_of_week')['order_id'].count()
day_names = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
print("\nOrders by Day of Week:")
for i, count in weekly_orders.items():
    print(f"  {day_names[i]}: {count}")

# Year-over-year (if multiple years)
yearly_orders = master_df.groupby('order_year')['order_id'].count()
print(f"\nYearly Orders:")
print(yearly_orders)

# ============================================
# FINAL SUMMARY
# ============================================
print("\n" + "="*50)
print("FINAL SUMMARY STATISTICS")
print("="*50)

print(f"Total Orders: {master_df['order_id'].nunique():,}")
print(f"Total Revenue: ${master_df['total_payment_value'].sum():,.2f}")
print(f"Average Order Value: ${master_df['total_payment_value'].mean():,.2f}")
print(f"Total Customers: {master_df['customer_id'].nunique():,}")
print(f"Average Products per Order: {master_df['total_products'].mean():.2f}")
print(f"Orders with Multiple Products: {master_df['is_multi_product'].sum():,} ({master_df['is_multi_product'].mean()*100:.1f}%)")
print(f"Total Shipping Revenue: ${master_df['total_shipping'].sum():,.2f}")
print(f"Average Shipping per Order: ${master_df['total_shipping'].mean():,.2f}")


CUSTOMER ANALYSIS

Top 10 Customers by Total Spent:
              total_orders  total_spent  total_revenue  customer_order_count
customer_id                                                                 
bmv1Lg65SUWP             1      7274.88         231.80                     1
Nz66V2TeAxAX             1      6929.31        1215.11                     1
H1deBLRJV7r9             1      6922.21         125.21                     1
oW2VygXvCSzO             1      6922.21         509.03                     1
r6BPYyzgP2BJ             1      6726.66         412.63                     1
CPjSUsnYD1a2             1      6081.54         139.97                     1
Cvkab0E6JBdW             1      6081.54         139.97                     1
sCMzIYs0fdq8             1      6081.54         139.97                     1
CPhZnWwAmVFJ             1      4950.34          56.15                     1
D8Jv4fSRxAar             1      4950.34         180.29                     1

DELIVERY ANALYSIS
On-ti

In [10]:
# Check if any orders have multiple products
multi_product_check = master_df[master_df['total_products'] > 1]
print(f"Orders with multiple products: {len(multi_product_check)}")
print(f"Sample of multi-product orders:")
print(multi_product_check[['order_id', 'total_products', 'product_categories']].head())

Orders with multiple products: 0
Sample of multi-product orders:
Empty DataFrame
Columns: [order_id, total_products, product_categories]
Index: []


In [12]:
# Check raw order_items data
print("="*50)
print("RAW ORDER_ITEMS DATA CHECK")
print("="*50)

# Count products per order in the raw order_items table
raw_product_count = pd.read_sql("""
    SELECT TOP 10 order_id, COUNT(*) AS product_count
    FROM order_items
    GROUP BY order_id
    HAVING COUNT(*) > 1
    ORDER BY product_count DESC
""", engine)

print(f"Orders with multiple products in raw data: {len(raw_product_count)}")
print(raw_product_count.head(10))

# Check a specific order that should have multiple products
sample_check = pd.read_sql("""
    SELECT TOP 10 o.order_id, COUNT(oi.product_id) AS item_count
    FROM orders o
    LEFT JOIN order_items oi ON o.order_id = oi.order_id
    GROUP BY o.order_id
    ORDER BY item_count DESC
""", engine)

print("\nSample of orders with their product counts:")
print(sample_check)

# Check if there are ANY orders with multiple products
multi_check = pd.read_sql("""
    SELECT COUNT(*) AS multi_product_orders
    FROM (
        SELECT order_id
        FROM order_items
        GROUP BY order_id
        HAVING COUNT(*) > 1
    ) AS multi
""", engine)

print(f"\nTotal orders with multiple products: {multi_check['multi_product_orders'].iloc[0]}")

# Check total rows in order_items
total_items = pd.read_sql("SELECT COUNT(*) AS total_rows FROM order_items", engine)
print(f"Total rows in order_items: {total_items['total_rows'].iloc[0]}")

# Check total unique orders
unique_orders = pd.read_sql("SELECT COUNT(DISTINCT order_id) AS unique_orders FROM order_items", engine)
print(f"Unique orders in order_items: {unique_orders['unique_orders'].iloc[0]}")

RAW ORDER_ITEMS DATA CHECK
Orders with multiple products in raw data: 0
Empty DataFrame
Columns: [order_id, product_count]
Index: []

Sample of orders with their product counts:
       order_id  item_count
0  00ejn2U1DdAv           1
1  00e8DgceulJY           1
2  00Dwp46qI5I4           1
3  00cXu09Yoyls           1
4  00C1YjVT8IOW           1
5  007kNCWr6IoE           1
6  0041NijdAYB0           1
7  002ohHsWSKrP           1
8  002Jk8hp9Bhy           1
9  0029bAIZFMKA           1

Total orders with multiple products: 0
Total rows in order_items: 89316
Unique orders in order_items: 89316


In [13]:
master_df.to_csv('ecommerce_dashboard_clean.csv', index=False)
print("Saved final dashboard data")

Saved final dashboard data


In [14]:
# First, make sure master_df has the correct data types
master_df_clean = master_df.copy()

# Convert date columns to string format that Power BI recognizes
date_columns = ['order_purchase_timestamp', 'order_approved_at', 
                'order_delivered_timestamp', 'order_estimated_delivery_date']

for col in date_columns:
    if col in master_df_clean.columns:
        master_df_clean[col] = pd.to_datetime(master_df_clean[col], errors='coerce')
        master_df_clean[col] = master_df_clean[col].dt.strftime('%Y-%m-%d %H:%M:%S')

# Replace NaN/None with appropriate values
master_df_clean = master_df_clean.fillna({
    'delivery_time_days': 0,
    'delivery_delay': 0,
    'is_late': 0,
    'product_categories': 'Unknown',
    'payment_types': 'Unknown'
})

# Ensure numeric columns are proper numbers
numeric_cols = ['total_products', 'total_price', 'total_shipping', 'payment_count',
                'total_payment_value', 'max_installments', 'delivery_time_days',
                'estimated_delivery_days', 'delivery_time', 'estimated_delivery_time',
                'delivery_delay', 'total_order_value', 'avg_item_price', 'order_year',
                'order_month', 'order_day', 'order_hour', 'order_day_of_week',
                'customer_order_count', 'avg_payment_per_installment', 'shipping_ratio']

for col in numeric_cols:
    if col in master_df_clean.columns:
        master_df_clean[col] = pd.to_numeric(master_df_clean[col], errors='coerce').fillna(0)

# Ensure boolean columns are integers
bool_cols = ['is_late', 'is_delivered', 'order_weekend', 'is_multi_product']
for col in bool_cols:
    if col in master_df_clean.columns:
        master_df_clean[col] = master_df_clean[col].astype(int)

# Remove any special characters from text columns
text_cols = ['order_id', 'order_status', 'customer_id', 'customer_city', 
             'customer_state', 'product_categories', 'payment_types', 'order_month_name']

for col in text_cols:
    if col in master_df_clean.columns:
        master_df_clean[col] = master_df_clean[col].astype(str).str.replace('[^\w\s,/-]', '', regex=True)

# Save with UTF-8 encoding (no BOM to avoid Power BI issues)
master_df_clean.to_csv('ecommerce_dashboard_clean.csv', index=False, encoding='utf-8')
print(" Saved cleaned CSV for Power BI")

✅ Saved cleaned CSV for Power BI


In [15]:
# Save as Excel instead (Power BI handles Excel very well)
master_df_clean.to_excel('ecommerce_dashboard_clean.xlsx', index=False)
print(" Saved as Excel file")

✅ Saved as Excel file


In [16]:
# Just save your existing master_df with the right settings
master_df.to_csv('ecommerce_dashboard.csv', 
                 index=False, 
                 encoding='utf-8-sig')  # This encoding works better with Power BI

print("CSV saved - try loading this in Power BI")

CSV saved - try loading this in Power BI


In [18]:
# Fill the empty values in those columns
master_df_clean = master_df.copy()

# Fill empty avg_payment_per_installment with 0
master_df_clean['avg_payment_per_installment'] = master_df_clean['avg_payment_per_installment'].fillna(0)

# Fill empty shipping_ratio with 0
master_df_clean['shipping_ratio'] = master_df_clean['shipping_ratio'].fillna(0)

# Fill any other potential empty numeric columns
numeric_cols = master_df_clean.select_dtypes(include=['number']).columns
for col in numeric_cols:
    master_df_clean[col] = master_df_clean[col].fillna(0)

# Fill empty text columns with 'Unknown'
text_cols = master_df_clean.select_dtypes(include=['object']).columns
for col in text_cols:
    master_df_clean[col] = master_df_clean[col].fillna('Unknown')

# Save with utf-8-sig encoding
master_df_clean.to_csv('ecommerce_dashboard_fixed.csv', index=False, encoding='utf-8-sig')

print("Fixed CSV saved - try loading this in Power BI")
print(f"Total rows: {len(master_df_clean)}")
print(f"Total columns: {len(master_df_clean.columns)}")

Fixed CSV saved - try loading this in Power BI
Total rows: 89316
Total columns: 37


In [4]:
master_df.to_csv('../data/processed/ecommerce_dashboard_fixed.csv', index=False)